src/ddsrt/include/dds/ddsrt/events.h

```diff
@@ -0,0 +1,179 @@
+ * http://www.eclipse.org/legal/epl-2.0, or the Eclipse Distribution License
+ * v. 1.0 which is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR BSD-3-Clause
+ */
+#ifndef DDSRT_EVENTS_H
+#define DDSRT_EVENTS_H
+
+#include "dds/ddsrt/retcode.h"
+#include "dds/ddsrt/sockets.h"
+#include "dds/ddsrt/atomics.h"
+
+#if __APPLE__
+#include "dds/ddsrt/events/darwin.h"
+#else
+#include "dds/ddsrt/events/posix.h"
+#endif
+
+#if defined (__cplusplus)
+extern "C" {
+#endif
+
+  /**
+  * @name Definitions for event trigger flags.
+  * These flags describe the different type of trigger events which can be watched for.
+  */
+///@{
+  /**< the flag is not set */
+#define DDSRT_EVENT_FLAG_UNSET 0u
+  /**< resource has data available for reading */
+#define DDSRT_EVENT_FLAG_READ (1u<<0) 
+  /**< resource can be written to */
+//#define DDSRT_EVENT_FLAG_WRITE (1u<<1) 
+  /**< resource has been opened */
+//#define DDSRT_EVENT_FLAG_OPEN (1u<<2) 
+  /**< resource has been closed */
+//#define DDSRT_EVENT_FLAG_CLOSE (1u<<3) 
+  /**< timeout has occurred on the resource */
+//#define DDSRT_EVENT_FLAG_TIMEOUT (1u<<4)
+  /**< notification of ip address change on the resource*/
+//#define DDSRT_EVENT_FLAG_IP_CHANGE (1u<<5)  
+///@}
+
+  /**
+  * @brief Describes the type of object being monitored for events.
+  */
+  enum ddsrt_event_type {
+    ddsrt_event_type_unset, /**< unitialized state*/
+    ddsrt_event_type_socket /**< indicating a socket type connection*/
+    /*,
+    ddsrt_event_type_file,
+    ddsrt_event_type_ifaddr*/
+  };
+  typedef enum ddsrt_event_type ddsrt_event_type_t;
+
+  /**
+  * @brief Structure containing the information of which object is being watched.
+  *
+  * Is used by ddsrt_event_queue to keep track of objects being watched.
+  */
+  struct ddsrt_event {
+    uint32_t flags; /**< Type of event being watched for, composited from DDSRT_EVENT_FLAG_$EVENT_TYPE$.*/
+    ddsrt_event_type_t type; /**< Type of object being watched.*/
+    ddsrt_atomic_uint32_t triggered; /**< Trigger status after a call to ddsrt_event_queue_wait.*/
+    union {
+      struct { ddsrt_socket_t sock; } socket;
+      /*struct ddsrt_event_file file;  this is for future expansion to be able to register events on files*/
+      /*struct ddsrt_event_ifaddr ifaddr; this is for future expansion to be able to register IP address changes*/
+    } data; /**< Container for the object being watched.*/
+  };
+
+  typedef struct ddsrt_event ddsrt_event_t;
+
+  typedef struct ddsrt_event_queue ddsrt_event_queue_t;
+  
+  /**
+  * @brief Initializes an existing event to indicate a socket.
+  *
+  * @param[out] ev Pointer to the event to initialize.
+  * @param[in] sock Socket to initialize the event with.
+  * @param[in] flags Flags to set for the event.
+  *
+  * @returns DDS_RETCODE_OK if everything went OK.
+  */
+  dds_return_t ddsrt_event_socket_init(ddsrt_event_t* ev, ddsrt_socket_t sock, uint32_t flags);
```

Comment:
Presumably `ev` must not be a null pointer, adding a `ddsrt_nonnull` or `ddsrt_nonnull_all` is useful documentation — and sometimes it even makes the compilers produce useful warnings at the call sites! See https://github.com/eclipse-cyclonedds/cyclonedds/blob/master/src/ddsrt/include/dds/ddsrt/attributes.h

Reply:
Thanks for the suggestion!
I was using lots of checking whether pointers were not null with assert, which seemed suboptimal in my eyes.
Will examine the use of ddsrt_nonnull and how to use it.

Reply:
This is now implemented.